File: bumperbot_vision/bumperbot_vision/bag_of_words.py
#!/usr/bin/env python3

import os
import rclpy
from rclpy.node import Node

import cv2
import numpy as np
from cv_bridge import CvBridge, CvBridgeError

import message_filters
from sensor_msgs.msg import Image, CameraInfo
from geometry_msgs.msg import TransformStamped

from std_srvs.srv import Trigger
from bumperbot_msgs.srv import SetString

from tf2_ros import Buffer, TransformListener, TransformBroadcaster
from tf2_ros import TransformException


class BagOfWords(Node):
    def __init__(self, name="bag_of_words"):
        super().__init__(name)

        self.is_training = True
        self.is_trained = False

        self.declare_parameter("training_frequency", 10.0)
        self.declare_parameter("vocabulary_size", 100)
        
        self.training_frequency = self.get_parameter("training_frequency").value
        self.vocabulary_size = self.get_parameter("vocabulary_size").value
        self.last_train_time = self.get_clock().now()

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.transform_broadcaster = TransformBroadcaster(self)

        self.bridge = CvBridge()

        self.image_sub = message_filters.Subscriber(self, Image, "left_camera/image_raw")
        self.info_sub = message_filters.Subscriber(self, CameraInfo, "left_camera/camera_info")
        
        self.approx_sync = message_filters.ApproximateTimeSynchronizer(
            [self.image_sub, self.info_sub], queue_size=10, slop=0.1)
        self.approx_sync.registerCallback(self.image_callback)

        self.detector = cv2.ORB_create()
        
        # We must use KDTree (algorithm=1) because KMeans vocabulary is float32
        index_params = dict(algorithm=1, trees=5)
        search_params = dict(checks=50)
        self.matcher = cv2.FlannBasedMatcher(index_params, search_params)
        
        # Data storage
        self.training_descriptors = [] 
        self.stored_poses = []
        self.database_histograms = []
        self.vocabulary = None

        self.train_mode_srv = self.create_service(Trigger, f"{name}/train_mode", self.train_mode_cb)
        self.test_mode_srv = self.create_service(Trigger, f"{name}/test_mode", self.test_mode_cb)
        self.save_srv = self.create_service(SetString, f"{name}/save", self.save_cb)
        self.load_srv = self.create_service(SetString, f"{name}/load", self.load_cb)


    def train_mode_cb(self, request, response):
        self.is_training = True
        response.success = True
        response.message = "Switched to train mode"
        return response


    def test_mode_cb(self, request, response):
        self.is_training = False
        response.success = True
        response.message = "Switched to test mode"
        return response


    def _compute_histogram(self, descriptors):
        """Manually matches descriptors to the vocabulary and builds a normalized histogram."""
        if self.vocabulary is None:
            return None
            
        # ORB outputs uint8, but vocabulary from KMeans is float32. We must cast to match.
        desc_float = np.float32(descriptors)
        
        # Match each descriptor to the nearest cluster center (visual word)
        matches = self.matcher.match(desc_float, self.vocabulary)
        
        # Build the histogram
        hist = np.zeros(self.vocabulary_size, dtype=np.float32)
        for m in matches:
            hist[m.trainIdx] += 1.0
            
        # Normalize
        hist_sum = np.sum(hist)
        if hist_sum > 0:
            hist /= hist_sum
            
        return hist


    def build_vocabulary_and_database(self):
        if not self.training_descriptors:
            self.get_logger().warning("No features gathered! Cannot build vocabulary.")
            return False

        self.get_logger().info("Clustering visual words... This may take a moment.")
        
        # Train the vocabulary using all gathered descriptors
        bow_train = cv2.BOWKMeansTrainer(self.vocabulary_size)
        for desc in self.training_descriptors:
            bow_train.add(np.float32(desc))
            
        self.vocabulary = bow_train.cluster()
        
        # Build the database histograms manually
        self.database_histograms = []
        for desc in self.training_descriptors:
            hist = self._compute_histogram(desc)
            self.database_histograms.append(hist)

        self.is_trained = True
        self.get_logger().info("Vocabulary and Database successfully built!")
        return True


    def save_cb(self, request, response):
        path = request.data
        if not path or not os.path.exists(path) or not os.path.isdir(path):
            response.success = False
            response.message = "Path cannot be empty or invalid"
            return response

        if not self.is_trained:
            if not self.build_vocabulary_and_database():
                response.success = False
                response.message = "Failed to build vocabulary. Cannot save database."
                rclpy.error(self.get_logger(), response.message)
                return response

        try:
            if self.vocabulary is None:
                raise RuntimeError("No vocabulary built to save. Switch to test mode first.")
                
            np.save(os.path.join(path, "bow_vocabulary.npy"), self.vocabulary)
            np.save(os.path.join(path, "bow_database.npy"), np.array(self.database_histograms))

            with open(os.path.join(path, "bow_poses.txt"), 'w') as f:
                for pose in self.stored_poses:
                    t = pose.transform.translation
                    r = pose.transform.rotation
                    f.write(f"{t.x} {t.y} {t.z} {r.x} {r.y} {r.z} {r.w}\n")

            response.success = True
            response.message = f"Database saved to {path}"
        except Exception as e:
            response.success = False
            response.message = f"Failed to save database: {str(e)}"
            
        return response


    def load_cb(self, request, response):
        path = request.data
        if not path or not os.path.exists(path) or not os.path.isdir(path):
            response.success = False
            response.message = "Path cannot be empty or invalid"
            return response

        try:
            self.vocabulary = np.load(os.path.join(path, "bow_vocabulary.npy"))
            db_hists = np.load(os.path.join(path, "bow_database.npy"))
            self.database_histograms = list(db_hists)

            poses_file = os.path.join(path, "bow_poses.txt")
            self.stored_poses.clear()
            with open(poses_file, 'r') as f:
                for line in f:
                    parts = line.strip().split()
                    if len(parts) == 7:
                        pose = TransformStamped()
                        pose.header.frame_id = "map"
                        pose.child_frame_id = "base_footprint"
                        pose.transform.translation.x = float(parts[0])
                        pose.transform.translation.y = float(parts[1])
                        pose.transform.translation.z = float(parts[2])
                        pose.transform.rotation.x = float(parts[3])
                        pose.transform.rotation.y = float(parts[4])
                        pose.transform.rotation.z = float(parts[5])
                        pose.transform.rotation.w = float(parts[6])
                        self.stored_poses.append(pose)

            self.is_trained = True
            self.is_training = False
            response.success = True
            response.message = f"Database loaded from {path}"
        except Exception as e:
            response.success = False
            response.message = f"Failed to load database: {str(e)}"
            rclpy.error(self.get_logger(), response.message)
            
        return response


    def image_callback(self, image_msg, info_msg):
        try:
            img = self.bridge.imgmsg_to_cv2(image_msg, desired_encoding="mono8")
        except CvBridgeError as e:
            self.get_logger().error(f"cv_bridge exception: {e}")
            return

        keypoints, descriptors = self.detector.detectAndCompute(img, None)

        if descriptors is None or len(descriptors) == 0:
            return

        if self.is_training:
            now = self.get_clock().now()
            time_diff = (now - self.last_train_time).nanoseconds / 1e9
            
            if time_diff < (1.0 / self.training_frequency):
                return
            
            self.last_train_time = now

            try:
                current_pose = self.tf_buffer.lookup_transform(
                    "map", "base_footprint", rclpy.time.Time())
            except TransformException as ex:
                self.get_logger().warning(
                    f"Could not transform: {ex}", throttle_duration_sec=2.0)
                return

            self.training_descriptors.append(descriptors)
            self.stored_poses.append(current_pose)
            self.get_logger().info(
                f"Added image to training set. Total images: {len(self.training_descriptors)}")
                
        else:
            if not self.is_trained:
                if not self.build_vocabulary_and_database():
                    return

            # Compute histogram using our custom matcher logic
            query_hist = self._compute_histogram(descriptors)
            
            if query_hist is not None and self.database_histograms:
                best_score = float('inf')
                best_id = -1
                
                for i, db_hist in enumerate(self.database_histograms):
                    score = cv2.norm(query_hist, db_hist, cv2.NORM_L1)
                    if score < best_score:
                        best_score = score
                        best_id = i

                if best_id != -1 and best_score < 1.5:  
                    pose = self.stored_poses[best_id]
                    pose.child_frame_id = "base_footprint_bow"
                    pose.header.stamp = self.get_clock().now().to_msg()
                    self.transform_broadcaster.sendTransform(pose)

                    self.get_logger().info(
                        f"Relocalized! Match Score: {best_score:.3f} | "
                        f"Pose [x: {pose.transform.translation.x:.2f}, y: {pose.transform.translation.y:.2f}]"
                    )


def main(args=None):
    rclpy.init(args=args)
    node = BagOfWords("bag_of_words")
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: bumperbot_vision/src/bag_of_words.cpp
#include <filesystem>

#include "bumperbot_vision/bag_of_words.hpp"

#include "cv_bridge/cv_bridge.hpp"

using namespace std::placeholders;

namespace bumperbot_vision
{
BagOfWords::BagOfWords(const std::string & name)
: Node(name), is_training_(true), is_trained_(false)
{
  approximate_sync_ = std::make_shared<ApproximateSync>(
    ApproximatePolicy(10), image_sub_, info_sub_);
  approximate_sync_->registerCallback(
    std::bind(&BagOfWords::imageCallback, this, _1, _2));
  image_transport::TransportHints hints{this};
  const auto sensor_data_qos = rclcpp::SensorDataQoS().get_rmw_qos_profile();

  training_frequency_ = declare_parameter("training_frequency", 10.0);
  last_train_time_ = now();

  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(get_clock());
  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
  transform_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(*this);

  image_sub_.subscribe(
    this, "left_camera/image_raw", hints.getTransport(), sensor_data_qos);
  info_sub_.subscribe(this, "left_camera/camera_info", sensor_data_qos);

  detector_ = cv::ORB::create();

  train_mode_srv_ = create_service<std_srvs::srv::Trigger>(
    name + "/train_mode", [&](const std_srvs::srv::Trigger::Request::SharedPtr request,
      std_srvs::srv::Trigger::Response::SharedPtr response) {
      (void)request;
      is_training_ = true;
      response->success = true;
      response->message = "Switched to train mode";
    });

  test_mode_srv_ = create_service<std_srvs::srv::Trigger>(
    name + "/test_mode", [&](const std_srvs::srv::Trigger::Request::SharedPtr request,
      std_srvs::srv::Trigger::Response::SharedPtr response) {
      (void)request;
      is_training_ = false;
      response->success = true;
      response->message = "Switched to test mode";
    });

  save_srv_ = create_service<bumperbot_msgs::srv::SetString>(
    name + "/save", [&](const bumperbot_msgs::srv::SetString::Request::SharedPtr request,
      bumperbot_msgs::srv::SetString::Response::SharedPtr response) {
      std::string path = request->data;
      if (path.empty() || !std::filesystem::exists(path) || !std::filesystem::is_directory(path)) {
        response->success = false;
        response->message = "Path cannot be empty or invalid";
        return;
      }

      if (!is_trained_) {
        if(!buildVocabulary()){
          response->success = false;
          response->message = "Failed to build vocabulary. Cannot save database.";
          RCLCPP_ERROR(get_logger(), response->message.c_str());
          return;
        }
      }
      voc_.save(path + "/bow_voc.bin");
      
      // Save the database in binary format
      std::ofstream feat_file(path + "/bow_features.bin", std::ios::binary);
      if (!feat_file.is_open()) throw std::runtime_error("Could not open bow_features.bin");

      size_t num_images = training_features_.size();
      feat_file.write(reinterpret_cast<const char*>(&num_images), sizeof(num_images));

      for(const auto& mat : training_features_) {
        int rows = mat.rows;
        int cols = mat.cols;
        int type = mat.type();
        feat_file.write(reinterpret_cast<const char*>(&rows), sizeof(rows));
        feat_file.write(reinterpret_cast<const char*>(&cols), sizeof(cols));
        feat_file.write(reinterpret_cast<const char*>(&type), sizeof(type));
          
        if (mat.isContinuous()) {
          feat_file.write(reinterpret_cast<const char*>(mat.ptr()), mat.total() * mat.elemSize());
        } else {
          cv::Mat cont = mat.clone();
          feat_file.write(reinterpret_cast<const char*>(cont.ptr()), cont.total() * cont.elemSize());
        }
      }
      feat_file.close();

      // Save the corresponding poses
      std::ofstream pose_file(path + "/bow_poses.txt");
      for (const auto& pose : stored_poses_) {
        pose_file << pose.transform.translation.x << " "
                  << pose.transform.translation.y << " "
                  << pose.transform.translation.z << " "
                  << pose.transform.rotation.x << " "
                  << pose.transform.rotation.y << " "
                  << pose.transform.rotation.z << " "
                  << pose.transform.rotation.w << "\n";
      }
      pose_file.close();

      response->success = true;
      response->message = "Database saved to " + path;
    });

  load_srv_ = create_service<bumperbot_msgs::srv::SetString>(
    name + "/load", [&](const bumperbot_msgs::srv::SetString::Request::SharedPtr request,
      bumperbot_msgs::srv::SetString::Response::SharedPtr response) {
      std::string path = request->data;
      if (path.empty() || !std::filesystem::exists(path) || !std::filesystem::is_directory(path)) {
        response->success = false;
        response->message = "Path cannot be empty or invalid";
        return;
      }
      try {
        voc_.load(path + "/bow_voc.bin");
        
        // Load the binary features 
        std::ifstream feat_file(path + "/bow_features.bin", std::ios::binary);
        size_t num_images;
        feat_file.read(reinterpret_cast<char*>(&num_images), sizeof(num_images));
        training_features_.clear();
        
        for(size_t i = 0; i < num_images; ++i) {
          int rows, cols, type;
          feat_file.read(reinterpret_cast<char*>(&rows), sizeof(rows));
          feat_file.read(reinterpret_cast<char*>(&cols), sizeof(cols));
          feat_file.read(reinterpret_cast<char*>(&type), sizeof(type));
          
          cv::Mat mat(rows, cols, type);
          feat_file.read(reinterpret_cast<char*>(mat.data), mat.total() * mat.elemSize());
          training_features_.push_back(mat);
        }

        feat_file.close();
        db_.setVocabulary(voc_, false, 0);
        for(const auto& descriptors : training_features_) {
          db_.add(descriptors);
        }

        std::ifstream pose_file(path + "/bow_poses.txt");
        if (!pose_file.is_open()) {
          throw std::runtime_error("Poses file not found!");
        }

        // Restore poses
        stored_poses_.clear();
        double tx, ty, tz, rx, ry, rz, rw;
        while (pose_file >> tx >> ty >> tz >> rx >> ry >> rz >> rw) {
          geometry_msgs::msg::TransformStamped pose;
          pose.header.frame_id = "map";
          pose.child_frame_id = "base_footprint";
          pose.transform.translation.x = tx;
          pose.transform.translation.y = ty;
          pose.transform.translation.z = tz;
          pose.transform.rotation.x = rx;
          pose.transform.rotation.y = ry;
          pose.transform.rotation.z = rz;
          pose.transform.rotation.w = rw;
          stored_poses_.push_back(pose);
        }
        pose_file.close();

        is_trained_ = true;
        is_training_ = false;
        response->success = true;
        response->message = "Database loaded from " + path;
      } catch (const std::exception & e) {
        response->success = false;
        response->message = "Failed to load database: " + std::string(e.what());
        RCLCPP_ERROR(get_logger(), response->message.c_str());
      }
    });
}

void BagOfWords::imageCallback(
  const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info_msg)
{
  cv::Mat img;
  try {
    img = cv_bridge::toCvCopy(image_msg, image_msg->encoding)->image;
  } catch (cv_bridge::Exception& e) {
    RCLCPP_ERROR(get_logger(), "cv_bridge exception: %s", e.what());
    return;
  }

  // Detect and compute ORB features
  std::vector<cv::KeyPoint> keypoints;
  cv::Mat descriptors;
  detector_->detectAndCompute(img, cv::noArray(), keypoints, descriptors);

  if (descriptors.empty()) {
    return;
  }

  if (is_training_) {
    // Control training frequency
    if ((now() - last_train_time_).seconds() < (1.0 / training_frequency_)) {
      return;
    }
    last_train_time_ = now();

    // Get the pose of the robot at the time of the image
    geometry_msgs::msg::TransformStamped current_pose;
    try {
      current_pose = tf_buffer_->lookupTransform(
        "map", "base_footprint", tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000, 
                           "Could not transform: %s", ex.what());
      return; 
    }

    // Store features and pose for database creation
    training_features_.push_back(descriptors);
    stored_poses_.push_back(current_pose);
    RCLCPP_INFO(get_logger(), "Added image to training set. Total images: %zu", training_features_.size());
  } else {
    if (!is_trained_) {
      if (!buildVocabulary()){
        RCLCPP_WARN(get_logger(), "Not trained yet! Cannot perform localization.");
        return;
      }
    }

    // Query the database to find the most similar image
    DBoW3::QueryResults results;
    db_.query(descriptors, results, 1);

    if (!results.empty()) {
      auto best_match = results[0];
      int best_id = best_match.Id;

      if (best_id >= 0 && best_id < static_cast<int>(stored_poses_.size())) {
        auto pose = stored_poses_[best_id];
        pose.child_frame_id = "base_footprint_bow";
        pose.header.stamp = now();
        transform_broadcaster_->sendTransform(pose);
        
        RCLCPP_INFO(get_logger(), 
          "Relocalized! Match Score: %.3f | Pose [x: %.2f, y: %.2f]",
          best_match.Score, 
          pose.transform.translation.x, 
          pose.transform.translation.y);
      }
    }
  }
}

bool BagOfWords::buildVocabulary()
{
  if (training_features_.empty()) {
    RCLCPP_WARN(get_logger(), "No features gathered! Cannot build vocabulary.");
    return false;
  }
  RCLCPP_INFO(get_logger(), "Clustering visual words... This may take a moment.");
  voc_.create(training_features_);
  db_.setVocabulary(voc_, false, 0);
  for (const auto& descriptors : training_features_) {
    db_.add(descriptors);
  }
  is_trained_ = true;
  return true;
}
}  // namespace bumperbot_vision

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<bumperbot_vision::BagOfWords>("bag_of_words"));
  rclcpp::shutdown();
  return 0;
}